test/e2e/performanceprofile/functests/7_performance_kubelet_node/cgroups.go

```diff
@@ -52,30 +76,436 @@ var _ = Describe("[performance] Cgroups and affinity", Ordered, func() {
+		}
+
+		chkOvsCgroupLoadBalance := func(node *corev1.Node) (string, error) {
+			cmd := []string{"cat", "/rootfs/sys/fs/cgroup/cpuset/ovs.slice/cpuset.sched_load_balance"}
+			return nodes.ExecCommandOnNode(cmd, workerRTNode)
+		}
+
+		DescribeTable("Verify cgroup layout on worker node",
+			func(crun bool, ovsProc, ovsCpus, ovslb bool) {
+				if crun {
+					By("Enabling crun")
+					mcpLabel := profilecomponent.GetMachineConfigLabel(profile)
+					key, value := components.GetFirstKeyAndValue(mcpLabel)
+					mcp, err := mcps.GetByLabel(key, value)
+					Expect(err).ToNot(HaveOccurred())
+					ctrcfg := mcps.EnableCrun("cnf-crun", profile, &mcp[0])
+					testclient.Client.Create(context.TODO(), ctrcfg)
+
+					By("Applying changes in performance profile and waiting until mcp will start updating")
+					mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+
+					By("Waiting when mcp finishes updates")
+					mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+
+					defer func() {
+						By("Disabling crun")
+						testclient.Client.Delete(context.TODO(), ctrcfg)
+						By("Applying changes in performance profile and waiting until mcp will start updating")
+						mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdating, corev1.ConditionTrue)
+						By("Waiting when mcp finishes updates")
+						mcps.WaitForCondition(performanceMCP, machineconfigv1.MachineConfigPoolUpdated, corev1.ConditionTrue)
+					}()
+				}
+
+				if ovsProc {
+					result, err := chkOvsCgrpProcs(workerRTNode)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(result).To(Not(BeEmpty()))
+				}
+
+				if ovsCpus {
+					result, err := chkOvsCgrpCpuset(workerRTNode)
+					ovsCPUSet, err := cpuset.Parse(result)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(ovsCPUSet).To(Equal(onlineCPUSet))
+				}
+				if ovslb {
+					result, err := chkOvsCgroupLoadBalance(workerRTNode)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(result).To(Equal("0"))
+				}
+			},
+			Entry("[test_id:68385] Verify ovs cgroup with crun ", true, true, true, true),
+			Entry("[test_id:64098] Verify ovs cgroup with runc", false, true, true, true),
+		)
+	})
+
+	Describe("Affinity", func() {
+		Context("ovn-kubenode Pods affinity ", func() {
+			testutils.CustomBeforeAll(func() {
+				initialProfile = profile.DeepCopy()
+			})
+			It("[test_id:64100] matches with ovs process affinity", func() {
+				ovnPod, err := getOvnPod(workerRTNode)
+				Expect(err).ToNot(HaveOccurred(), "Unable to get ovnPod")
+
+				ovnContainersids, err := getOvnPodContainers(&ovnPod)
+				Expect(err).ToNot(HaveOccurred())
+
+				// Generally there are many containers inside a kubenode pods
+				// we don't need to check cpus used by all the containers
+				// we take first container
+				cpus := getCpusUsedByOvnContainer(workerRTNode, ovnContainersids[0])
```

Comment:
There are more containers than just one in the OVN-K pod, are you sure the [0] will pick the right one?

Reply:
It probably works now, but there is a hidden assumption here that all are burstable.

Reply:
@MarSik We are picking the ovnkubenode-pod running on the worker-cnf node ,   there 6 containers in each ovnkubenode pod All the containers have the same the cpu affinity 

Example from worker-cnf node:

```
d9a47cd57d88b       254a65c3ca4ef4a15337e1ea6d9fcc009ae81ce11ad2385a7e57fcdcc00ab856   4 hours ago         Running             nbdb                                 11                  e9a21fe08292b       ovnkube-node-rx8ng
a02982b6a5c69       254a65c3ca4ef4a15337e1ea6d9fcc009ae81ce11ad2385a7e57fcdcc00ab856   4 hours ago         Running             northd                               11                  e9a21fe08292b       ovnkube-node-rx8ng
1f358afb3ca9a       39ba63a1ff406d50e8d3d0d67b4028d33be766e49d9fe0fcc77b5f3f2a24e51f   4 hours ago         Running             kube-rbac-proxy-ovn-metrics          11                  e9a21fe08292b       ovnkube-node-rx8ng
fe0c694d25dab       39ba63a1ff406d50e8d3d0d67b4028d33be766e49d9fe0fcc77b5f3f2a24e51f   4 hours ago         Running             kube-rbac-proxy-node                 11                  e9a21fe08292b       ovnkube-node-rx8ng
cac2002c520b2       254a65c3ca4ef4a15337e1ea6d9fcc009ae81ce11ad2385a7e57fcdcc00ab856   4 hours ago         Running             ovn-acl-logging                      11                  e9a21fe08292b       ovnkube-node-rx8ng
c9a851fc9db71       254a65c3ca4ef4a15337e1ea6d9fcc009ae81ce11ad2385a7e57fcdcc00ab856   4 hours ago         Running             ovn-controller                       11                  e9a21fe08292b       ovnkube-node-rx8ng
```

All the containers have same cpu affinity:

```[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]#  find /sys/fs/cgroup/cpuset -name crio-a02982b6a5c69*
/sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-a02982b6a5c69e3c67f763c8edb2afceede88cae47523853fedbed070b5884fc.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-a02982b6a5c69e3c67f763c8edb2afceede88cae47523853fedbed070b5884fc.scope/cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]#  find /sys/fs/cgroup/cpuset -name crio-1f358afb3ca9a*
/sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-1f358afb3ca9a32726b03c101bbd6e1825feba8c7c2ccdde6a64a1aa7c1106d6.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-1f358afb3ca9a32726b03c101bbd6e1825feba8c7c2ccdde6a64a1aa7c1106d6.scope/cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]#  find /sys/fs/cgroup/cpuset -name crio-fe0c694d25dab*
/sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-fe0c694d25dabe1eb9394a40242e1133673e5f3689f994cc1db464525a2047fb.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-fe0c694d25dabe1eb9394a40242e1133673e5f3689f994cc1db464525a2047fb.scope/cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]#  find /sys/fs/cgroup/cpuset -name crio-cac2002c520b2*
/sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-cac2002c520b2c2d0c84c17b21a3cb4224ef024dc9f3c195b3e00813b1e8d94e.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-cac2002c520b2c2d0c84c17b21a3cb4224ef024dc9f3c195b3e00813b1e8d94e.scope/cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]#  find /sys/fs/cgroup/cpuset -name crio-c9a851fc9db71*                                                                                                                          /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-c9a851fc9db71424838be9d5e8d6b9aa435bbb6db54e6bac1ea5733af038cfbb.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-c9a851fc9db71424838be9d5e8d6b9aa435bbb6db54e6bac1ea5733af038cfbb.scope/cpuset.cpus
0-3,5,7-23
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# find /sys/fs/cgroup/cpuset -name crio-d9a47cd57d88b*
/sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope
[root@ocp-worker-0 crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope]# cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod051b68ac_f4ab_4761_8979_5f6a9bf746e3.slice/crio-d9a47cd57d88b0114c63953bb95955db2c3797e059d71c0cb5aead177a332de9.scope/cpuset.cpus
0-3,5,7-23

```

Reply:
Right now they do, but as I said, it is a hidden assumption that this will not change.

Reply:
Okay do we have any other reliable method.

Reply:
Can we find the names of the containers and pick the right one?